### ISSUE TITLE
[TF 2.0 eager] tf.gather_nd gradient outputs incorrect result

### ISSUE THREAD

**Issue by ZhiqingXiao** [tensorflow/tensorflow#27526]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10 x64
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): tensorflow 2.0.0-alpha0
- Python version: 3.7.3
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: None
- GPU model and memory: None

You can collect some of this information using our environment capture [script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with
python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

**Describe the current behavior**

`tape.gradient` can not pass through indices or `tf.gather_nd` and returns `None`.

**Describe the expected behavior**

`tape.gradient` should be able to pass through indices or `tf.gather_nd` and returns meaning numerical results.

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

```python
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.optimizers import Adam

model = keras.Sequential()
model.add(keras.layers.Dense(units=3, activation=tf.nn.softmax))
model.compile(optimizer=Adam(0.1),
        loss=keras.losses.categorical_crossentropy)

x = tf.convert_to_tensor([[0.9, -0.8, ],], dtype=tf.float32)
with tf.GradientTape(persistent=True) as tape:
    y = model(x)
grad_tensors = tape.gradient(y, model.variables)
print(grad_tensors) # expected results printed
grad_tensors = tape.gradient(y[0, 0], model.variables)
print(grad_tensors) # print incorrect results
grad_tensors = tape.gradient(tf.gather_nd(y, [0, 0]), model.variables)
print(grad_tensors) # print incorrect results
```

*Current Output:*
The first line without gather outputs correct gradients. But the 2nd and 3rd line return None, that is incorrect. It should have returned an element within the first output.
```
[<tf.Tensor: id=209, shape=(2, 3), dtype=float32, numpy=
array([[0., 0., 0.],
       [0., 0., 0.]], dtype=float32)>, <tf.Tensor: id=207, shape=(3,), dtype=float32, numpy=array([0., 0., 0.], dtype=float32)>]
[None, None]
[None, None]
```

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.
